octree_creation_app/driver.py

```diff
@@ -72,50 +72,82 @@ def octree_from_params(params: OctreeParams):
         )
 
         for label, value in params.free_parameter_dict.items():
+
             refinement_object = getattr(params, value["object"])
-            levels = utils.str2list(getattr(params, value["levels"]))
             if not isinstance(refinement_object, ObjectBase):
                 continue
+            levels = utils.str2list(getattr(params, value["levels"]))
+
+            mesh = OctreeDriver.refine_by_object_type(
+                mesh=mesh,
+                refinement_object=refinement_object,
+                levels=levels,
+                horizon=getattr(params, value["horizon"]),
+                distance=getattr(params, value["distance"]),
+                diagonal_balance=params.diagonal_balance,
+            )
 
             print(f"Applying {label} on: {getattr(params, value['object']).name}")
 
-            if getattr(params, value["horizon"]):
-                mesh = OctreeDriver.refine_tree_from_surface(
-                    mesh,
-                    refinement_object,
-                    levels,
-                    params.diagonal_balance,
-                    max_distance=getattr(params, value["distance"]),
-                )
+        print("Finalizing . . .")
+        mesh.finalize()
+        octree = treemesh_2_octree(params.geoh5, mesh, name=params.ga_group_name)
 
```

Comment:
Would you mind adding a logging about being done?

Reply:
done.